Re-use identical tile's geometry and oriented the main normal on the Z axis

gchoqueux opened the pull request:
## Description
Re-use identical tile's geometry and oriented the main normal on the Z axis
* The normal in the center of the geometry is aligned with the Z axis.
* Bounding boxes can use the Parent's worldMatrix (tile).

The geometry normal is computed with correct geodesic normal.
The bounding boxes is calculated by THREE.js.

## Motivation and Context
* reduce memory
* increase performance
* work tile's space, by example wfs (fix #558 issue)
* simplify obb computation
* reduce/fix obb calculation errors see #87 issues
* fix obb/culling for pano tile

**WARNING**
change subdivision count in PLANAR test, the bounding sphere are bigger, so the subdivision are changed

gchoqueux committed:
feat(examples): add gui debug in planar in debug mode only

gchoqueux committed:
feat(core): add bounding sphere helper in debug mode

peppsac commented:
3 fixes, 2 feat, 1 refactor => please, split this PR.

gchoqueux commented:
@peppsac 
normally, these commits should be squashed, because  last 2 commits fix the first: 
* refactor (core): use common geometry in meshTile b5b20f2 
* fix (test): !change subdivision count in PLANAR test! 11d9e71
* fix (protocols): wfs, adapt the mesh transformation with the tile space e2e7c8b

and the 2 debugging features are needed to verify how this PR works :
* feat (examples): add gui debug in planar in debug mode only 70a6cb0 
* feat (core) : add bounding sphere helper in debug mode 65220e3

I remove 
fix (core): improve search for min/max altitude in xbil buffer  663b96d5

peppsac commented on index.html:
```diff
@@ -45,9 +45,9 @@
-            promises.push(itowns.Fetcher.json('examples/layers/JSONLayers/OrthosCRS.json').then(function(result) { return globeView.addLayer(result); }));
-            promises.push(itowns.Fetcher.json('examples/layers/JSONLayers/ScanEX.json').then(function(result) { return globeView.addLayer(result); }));
-            promises.push(itowns.Fetcher.json('examples/layers/JSONLayers/Region.json').then(function(result) { return globeView.addLayer(result); }));
+            // promises.push(itowns.Fetcher.json('examples/layers/JSONLayers/OrthosCRS.json').then(function(result) { return globeView.addLayer(result); }));
```
You forgot to remove this temporary change

gchoqueux replied:
ok

peppsac commented on src/Core/Geographic/Coordinates.js:
```diff
@@ -136,7 +136,8 @@ function _convert(coordsIn, newCrs, target) {
     } else {
         if (coordsIn.crs === 'EPSG:4326' && newCrs === 'EPSG:4978') {
-            const cartesian = ellipsoid.cartographicToCartesian(coordsIn);
+            target._normal = ellipsoid.geodeticSurfaceNormalCartographic(coordsIn);
```
Doing:
```js
target._normal = coordsIn.geodesicNormal;
```
Would avoid recomputing the normal if it was already done for `coordsIn`

gchoqueux replied:
done

peppsac commented on src/Core/Math/Ellipsoid.js:
```diff
@@ -56,10 +56,10 @@ Ellipsoid.prototype.setSize = function setSize(size) {
 
 
-Ellipsoid.prototype.cartographicToCartesian = function cartographicToCartesian(coordCarto) {
+Ellipsoid.prototype.cartographicToCartesian = function cartographicToCartesian(coordCarto, normal) {
```
You must change the code to not modify the `normal` given by the user (modifying parameters makes the API error-prone)

gchoqueux replied:
done

peppsac commented on src/Core/Prefab/GlobeView.js:
```diff
@@ -163,6 +164,31 @@ export function createGlobeLayer(id, options) {
+    // return common extent to pool the geometries
+    // the geometry in common extent is identical to the existing input
+    // with a transformation (translation, rotation)
+    wgs84TileLayer.getCommonGeometryExtent = (extent) => {
```
These functions don't belong to `GlobeView` because:
- the panorama layers will want to use them
- they are tile geometries implementation details

peppsac replied:
Also the naming is wrong: you're basically doing a modulo operation on the longitude component (why `sizeLatitude`?) of the extent so the name has to express this.
Btw, why aren't you doing the same thing for the other angle? (I guess it has to do with the ellipsoid shape, but you may want to add a comment in the code)

gchoqueux replied:
* `getCommonGeometryExtent` moved in tile builder
* `sizeLatitude` is modified
* `getCommonGeometryExtent`, in function of builder, it doesn't do modulo

peppsac commented on src/Core/Prefab/GlobeView.js:
```diff
@@ -163,6 +164,31 @@ export function createGlobeLayer(id, options) {
+
+    // return rotation to transform tile to position it on ellipsoid
+    // this transformation take into account the transformation of the parents
+    wgs84TileLayer.getQuaternionTile = (tile) => {
```
Wrong naming: the name doesn't express what the function does.
And the argument should be an extent, since you don't use any other properties of the `tile`.

gchoqueux replied:
changed in `getQuaternionFromExtent`

peppsac commented on src/Core/Prefab/Planar/PlanarTileBuilder.js:
```diff
@@ -50,19 +50,8 @@ PlanarTileBuilder.prototype.vProjecte = function vProjecte(v, params)
-    const translate = new THREE.Vector3(0, 0, 0);
-    // normal is up vector
-    return new OBB(min, max, undefined, translate);
+PlanarTileBuilder.prototype.OBB = function OBBFn(params, boundingBox) {
```
Why keep the `params` parameter if it's unused?

peppsac replied:
(same question for `BuilderEllipsoidTile`)

gchoqueux replied:
cleaned

peppsac commented on src/Core/Scheduler/Providers/TileProvider.js:
```diff
@@ -49,50 +41,84 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
+    const south = ce.south().toFixed(8);
+
+    // build geometry if doesn't exist
+    if (!this.cacheGeometry[level].get(south)) {
```
This is wrong: the `south` value alone can't serve as a cache key.
You must at least depend on `command.layer.disableSkirt` (and probably on ellipsoidSize - even if currently this cannot be changed) and the `builder` used.

gchoqueux replied:
yes, in another PR, there's not user cases for the moment

peppsac commented on src/Core/Scheduler/Providers/TileProvider.js:
```diff
@@ -49,50 +41,84 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
+
+    // build geometry if doesn't exist
+    if (!this.cacheGeometry[level].get(south)) {
+        this.cacheGeometry[level].set(south, defer());
```
Why is it necessary to use a Promise mechanism here (instead of storing the geometry directly in the cache)?

gchoqueux replied:
As the commands are asynchronous many tiles may ask for geometry during construction

autra replied:
Everything here is synchronous, so it is guaranteed that no other part of itowns (or any other js in this page really) will get executed between this line and line 83. So the deferred will actually never be used here and you can directly set the TileGeometry in the cacheGeometry, without all the promise boilerplate.

gchoqueux replied:
However the commands are executed in asynchronous, many tiles can ask for geometry during construction.
I tested it didn't work

autra replied:
Please be more specific. What does "during construction" mean?

>  I tested it didn't work

What did you test? What didn't work?

> However the commands are executed in asynchronous

The fact that commands execute asynchronously has nothing to do with the matter. You set a promise on line 76 and resolve it on line 83.  No code can execute between these 2 lines, whether synchronous or asynchronous. 

Have you removed the `then` logic on line 87 when you tested?

gchoqueux replied:
Thanks @autra , I started this PR a long time ago, where commands was concurrent.
I remove the promise in provider

peppsac commented on src/Core/TileMesh.js:
```diff
@@ -25,11 +25,10 @@ function TileMesh(geometry, params) {
-    this.normal = params.center.clone().normalize();
 
-    this.boundingSphereOffset = new THREE.Vector3();
+    this.obb = this.geometry.OBB.clone();
```
What's the point of `this.obb`?

gchoqueux replied:
obliged to clone because the DEM is different for each tile,

peppsac commented on src/Core/TileMesh.js:
```diff
@@ -43,14 +42,17 @@ function TileMesh(geometry, params) {
 
     this.material.setUuid(this.id);
+
+    this.invWorldQuaternion = new THREE.Quaternion();
```
Why do you need `invWorldQuaternion`?

peppsac commented on src/Renderer/ThreeExtended/Feature2Mesh.js:
```diff
@@ -219,13 +221,14 @@ function coordinateToPolygonExtruded(coordinates, properties, options) {
         const { contour, property } = extractFeature(coordinates, properties, id);
         // get altitude and extrude amount from properties
         const altitudeBottom = getAltitude(options, property);
+        minAltitude = Math.min(minAltitude, altitudeBottom);
```
What is this `minAltitude` stuff? And why is that necessary?

gchoqueux replied:
It's the minimum altitude of the vertex of the points. It's used to position the object3D on the surface normal

peppsac commented on examples/globe_wfs_extruded.js:
```diff
@@ -90,13 +92,37 @@ function acceptFeature(properties) {
+    },
+};
+
+globeView.addFrameRequester(scaler);
```
Nice to get this working on the globe!

peppsac reviewed (changes requested):
Here's a first round of comments on the changes proposed.

gchoqueux committed:
fix(protocols): wfs, adapt the mesh transformation with the tile space

fix #558 issue

peppsac commented on src/Core/Math/Ellipsoid.js:
```diff
@@ -56,10 +56,10 @@ Ellipsoid.prototype.setSize = function setSize(size) {
     // var n;
     var k = new THREE.Vector3();
-    var n = this.geodeticSurfaceNormalCartographic(coordCarto);
+    const n = normal ? normal.clone() : this.geodeticSurfaceNormalCartographic(coordCarto);
```
You could replace this with:
```js
const n = normal ? normal.clone() : coordCarto.geodesicNormal.clone();
```
Or even better, you could avoid changing the API and you'll get the same result by doing:
```js
const n = coordCarto.geodesicNormal.clone();
```

peppsac commented on src/Core/Prefab/Globe/BuilderEllipsoidTile.js:
```diff
@@ -79,9 +87,37 @@ BuilderEllipsoidTile.prototype.getUV_PM = function getUV_PM(params) {
+    // but we should be reverse UV on tile
+    const sizeLongitude = Math.abs(extent.west() - extent.east()) / 2;
+    const communExtent = new Extent(extent.crs(), -sizeLongitude, sizeLongitude, extent.south(), extent.north());
+    communExtent._internalStorageUnit = extent._internalStorageUnit;
```
`commun` is not an English word :)

peppsac commented on src/Core/Prefab/Globe/BuilderEllipsoidTile.js:
```diff
@@ -79,9 +87,37 @@ BuilderEllipsoidTile.prototype.getUV_PM = function getUV_PM(params) {
+// return common extent to pool the geometries
+// the geometry in common extent is identical to the existing input
+// with a transformation (translation, rotation)
+BuilderEllipsoidTile.prototype.getCommonGeometryExtent = function getCommonGeometryExtentFn(extent) {
```
I'd prefer to reserve `getXXX` for simple getter method.

I have more comment about this API, see layer.

peppsac commented on src/Core/Scheduler/Providers/TileProvider.js:
```diff
@@ -49,47 +42,78 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
+        this.cacheGeometry[level] = new Map();
+    }
 
+    const ce = builder.getCommonGeometryExtent(extent);
```
I don't think TileProvider needs to be aware of this: this could an internal feature of each tile builder.

But even without this, the API  could be simplified a lot if `getCommonGeometryExtent` was something that would return 2 things: the new extent and the transformation to apply.
And this would give you a better naming as well: something like: `computeSharableExtent`.

WDYT?

peppsac commented on src/Core/Scheduler/Providers/TileProvider.js:
```diff
@@ -49,47 +42,78 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
+    const ce = builder.getCommonGeometryExtent(extent);
+    const south = ce.south().toFixed(8);
+
+    let geometry = this.cacheGeometry[level].get(south);
```
I already commented about this: you can't use `south` as a cache key - and no it won't wait for another PR because it can break today.
Additional things to used for the key: builder type, segment count and probably disableSkirt.

Also I'd prefer a single cache than 1 cache per level.

Last but not least: when are you cleaning this cache?
If I zoom in a lot and zoom out, the geometries for high levels will never get disposed..

peppsac reviewed (changes requested):
Looks better, thanks.

You didn't change the cleanup code (see ObjectRemovalHelper.removeChildrenAndCleanup) so geometries are going to be disposed even if they are still in used...

Another question: this change is introducing a behavior change - not an API one but a behavior one (tiles' transformation is different now). So you need to document this somewhere in your commit message with a BREAKING CHANGE tag so that `Release Notes` can grab this change.

peppsac commented on src/Core/Prefab/Globe/BuilderEllipsoidTile.js:
```diff
@@ -33,24 +38,30 @@ BuilderEllipsoidTile.prototype.Prepare = function Prepare(params) {
-BuilderEllipsoidTile.prototype.Center = function Center(params) {
-    params.center = params.extent.center(this.tmp.coords[0])
+BuilderEllipsoidTile.prototype.Center = function Center(extent) {
+    this.tmp.coords[0]._normal = undefined;
```
Can you please move this to `Extent` (probably in Extent.center() method) or in `Coordinates` (in the Coordinates.set())?

peppsac commented on src/Core/Prefab/Panorama/PanoramaTileBuilder.js:
```diff
@@ -97,57 +102,55 @@ PanoramaTileBuilder.prototype.vProjecte = function vProjecte(v, params) {
+    return new OBB(boundingBox.min, boundingBox.max);
+};
+
+PanoramaTileBuilder.prototype.getCommonGeometryExtent = function getCommonGeometryExtentFn(extent) {
```
Is this still used?

peppsac commented on src/Core/Prefab/Panorama/PanoramaTileBuilder.js:
```diff
@@ -97,57 +102,55 @@ PanoramaTileBuilder.prototype.vProjecte = function vProjecte(v, params) {
+    const sharableExtent = new Extent(extent.crs(), -sizeLongitude, sizeLongitude, extent.south(), extent.north());
+    sharableExtent._internalStorageUnit = extent._internalStorageUnit;
+
+    // compute rotation to transform tile to position it on ellipsoid
```
The comment refers to an "ellipsoid" but there's no ellipsoid in this context.
Can you please fix the comment?

peppsac commented on src/Core/Scheduler/Providers/TileProvider.js:
```diff
@@ -49,47 +42,73 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
+    const south = sharableExtent.south().toFixed(8);
 
-    var parent = command.requester;
+    console.log(quaternion, position);
```
This should probably be removed :)

peppsac commented on src/Core/Scheduler/Providers/TileProvider.js:
```diff
@@ -49,47 +42,73 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
+        };
+
+        geometry = new TileGeometry(paramsGeometry, builder);
+        this.cacheGeometry[level].set(south, geometry);
```
`level` isn't a good cache key (see earlier comment https://github.com/iTowns/itowns/pull/586#discussion_r157729520)

peppsac commented on src/Core/Scheduler/Providers/TileProvider.js:
```diff
@@ -34,6 +26,7 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
 
     layer.level0Nodes = [];
     layer.onTileCreated = layer.onTileCreated || (() => {});
+    layer.object3d.invQuaternion = layer.object3d.quaternion.clone().inverse();
```
`invQuaternion` is never updated. And IMHO it's a bad idea to introduce this concept: just call inverse() when you need it.
Or maybe rewrite the code doing this to avoid requiring a matrix inversion (in worldToLocal) + 2 quaternion.inverse() call.

gchoqueux commented:
@peppsac updated

peppsac commented on src/Process/ObjectRemovalHelper.js:
```diff
@@ -8,7 +8,12 @@ export default {
         } else {
             if (obj.geometry) {
-                obj.geometry.dispose();
+                if (obj.geometry.count) {
```
This isn't a good idea since you're breaking geometry disposal for other types of geometries.

See other comment  in TileProvider.js

peppsac commented on src/Process/PanoramaTileProcessing.js:
```diff
@@ -1,3 +1,5 @@
+// import * as THREE from 'three';
```
Remove dead code please.

peppsac commented on src/Core/Scheduler/Providers/TileProvider.js:
```diff
@@ -49,47 +41,80 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
+
+        geometry = new TileGeometry(paramsGeometry, builder);
+        geometry._count = 0;
+        this.cacheGeometry.set(key, geometry);
```
You could do this:
```js
geometry.dispose = () => {
    geometry._count--;
    if (geometry._count == 0) {
       THREE.BufferGeometry.prototype.dispose.call(geometry);
       this.cacheGeometry.delete(key);
    }
}
```
And remove the `count` property. This way the ref-counting feature would be internal to TileProvider.

peppsac reviewed (commented):
Did you observe some measurable memory or perf gains with this changeset?

gchoqueux commented:
> Did you observe some measurable memory or perf gains with this changeset?

the processes time (to build tiles) and used memory (geometryTile) are divided by 3

peppsac commented on src/Core/Scheduler/Providers/TileProvider.js:
```diff
@@ -49,47 +41,79 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
-    var tile = new TileMesh(geometry, params);
-
-    tile.layer = command.layer.id;
+    geometry.count++;
```
Shouldn't it be `geometry._count++`?

peppsac commented on src/Core/Geographic/Extent.js:
```diff
@@ -245,6 +245,7 @@ Extent.prototype.center = function center(target) {
     let c;
     if (target) {
         Coordinates.call(target, this._crs, this._values[0], this._values[2]);
+        target._normal = undefined;
```
Resetting the `_normal` value should be done inside `Coordinates`.

peppsac commented on src/Core/Prefab/Globe/BuilderEllipsoidTile.js:
```diff
@@ -33,24 +40,27 @@ BuilderEllipsoidTile.prototype.Prepare = function Prepare(params) {
 
 // get position 3D cartesian
 BuilderEllipsoidTile.prototype.VertexPosition = function VertexPosition(params) {
+    this.tmp.coords[0]._normal = undefined;
```
Proposal: move this reset in `Coordinates.set()` and use it here:
```js
this.tmp.coords[0].set(
    'EPSG:4326',
    params.projected.longitudeRad,
    params.projected.latitudeRad);
```

peppsac reviewed (commented):
> the processes time (to build tiles) and used memory (geometryTile) are divided by 3

This info should be in one of the commits message - and maybe add explain you measured it.

The code is almost good, I just left a few comments to address.

gchoqueux commented:
> the processes time (to build tiles) and used memory (geometryTile) are divided by 3

I compare master and this PR with different calls from the `GlobeControls` API :
 * Total running time on `TileProvider.executeCommand`
  * Snapshots of memory occupied by all instances of  `TileGeometry`

peppsac commented on src/Core/Prefab/Globe/BuilderEllipsoidTile.js:
```diff
@@ -33,24 +40,30 @@ BuilderEllipsoidTile.prototype.Prepare = function Prepare(params) {
 
     this.tmp.coords[0].as('EPSG:4978', this.tmp.coords[1]).xyz(this.tmp.position);
-    this.tmp.normal.copy(this.tmp.position).normalize();
+    this.tmp.normal.copy(this.tmp.coords[1].geodesicNormal);
```
Btw, you could remove `this.tmp.normal` completely and replace VertexNormal code by `return this.tmp.coords[1].geodesicNormal`.

peppsac reviewed (commented):
Thanks for all the edits, the PR looks good now.

I left a small suggestion to simplify the code. And of course you need to rebase / edit commits properly before merging :)

gchoqueux committed:
refactor(core): use common geometry in meshTile

 * identical geometries are reused to build the tileMeshs
 * the bounding boxes is calculated by THREE.js
 * bounding boxes can use the Parent's worldMatrix (tile).

BREAKING CHANGE
The geodesic normal in the center of the tile's geometry is aligned with the mesh's local Z axis.
Previously, the tiles were not oriented. their local landmark had the same orientation as the world landmark. You must take into account the orientation of the tile if you add objects to this tile.

Fixes:
 * Ellipsoid tile: The tile's geometry normal is computed with correct geodesic normal
 * reduces obb calculation errors see #87 issues
 * errors obb/culling for pano tile

Performance
 * the processes time (to build tiles in TileProvider.executeCommand) and used memory (geometryTile) are divided by ~3

gchoqueux committed:
fix(test): !change subdivision count in PLANAR test!

BREAKING CHANGE
the bounding sphere are bigger, so the subdivision are changed

gchoqueux commented:
@peppsac  thanks for review
commit message updated

gchoqueux commented:
@peppsac this PR can be merged

peppsac reviewed (approved):
Ok this PR looks good now; thanks for the hard work!

peppsac commented:
@gchoqueux This PR broke the http://www.itowns-project.org/itowns/examples/multiglobe.html example.

Could you take a look at this please? (I can open a new issue if you prefer)
